palace/models/spaceoperator.cpp

```diff
@@ -200,400 +172,733 @@ void SpaceOperator::CheckBoundaryProperties()
   }
 }
 
-void SpaceOperator::PrintHeader()
+namespace
+{
+
+void PrintHeader(mfem::ParFiniteElementSpace &h1_fespace,
+                 mfem::ParFiniteElementSpace &nd_fespace,
+                 mfem::ParFiniteElementSpace &rt_fespace, bool &print_hdr)
 {
   if (print_hdr)
   {
-    Mpi::Print("\nConfiguring system matrices, number of global unknowns: {:d}\n",
-               nd_fespaces.GetFinestFESpace().GlobalTrueVSize());
+    Mpi::Print("\nAssembling system matrices, number of global unknowns:\n"
+               " H1: {:d}, ND: {:d}, RT: {:d}\n",
+               h1_fespace.GlobalTrueVSize(), nd_fespace.GlobalTrueVSize(),
+               rt_fespace.GlobalTrueVSize());
     print_hdr = false;
   }
 }
 
-std::unique_ptr<petsc::PetscParMatrix>
-SpaceOperator::GetSystemMatrixPetsc(SpaceOperator::OperatorType type, double omega,
-                                    mfem::Operator::DiagonalPolicy ess_diag, bool print)
+template <typename T1, typename T2, typename T3, typename T4>
+auto BuildOperator(mfem::ParFiniteElementSpace &fespace, T1 *df, T2 *f, T3 *dfb, T4 *fb,
+                   mfem::AssemblyLevel assembly_level, int skip_zeros,
+                   bool no_assembly = false)
 {
-  // Construct the frequency-dependent complex linear system matrix:
-  //                 A = K + iω C - ω² (Mr + i Mi) + A2(ω)
-  // or any one of its terms.
-  const int sdim = nd_fespaces.GetFinestFESpace().GetParMesh()->SpaceDimension();
-  SumMatrixCoefficient dfr(sdim), dfi(sdim), fr(sdim), fi(sdim), fbr(sdim), fbi(sdim);
+  auto a = std::make_unique<mfem::SymmetricBilinearForm>(&fespace);
+  if (df && !df->empty())
+  {
+    a->AddDomainIntegrator(new mfem::CurlCurlIntegrator(*df));
+  }
+  if (f && !f->empty())
+  {
+    a->AddDomainIntegrator(new mfem::MixedVectorMassIntegrator(*f));
+  }
+  if (dfb && !dfb->empty())
+  {
+    a->AddBoundaryIntegrator(new mfem::CurlCurlIntegrator(*dfb));
+  }
+  if (fb && !fb->empty())
+  {
+    a->AddBoundaryIntegrator(new mfem::MixedVectorMassIntegrator(*fb));
+  }
+  if (!no_assembly)
+  {
+    a->SetAssemblyLevel(assembly_level);
+    a->Assemble(skip_zeros);
+    a->Finalize(skip_zeros);
+  }
+  return a;
+}
+
+template <typename T1, typename T2>
+auto BuildAuxOperator(mfem::ParFiniteElementSpace &fespace, T1 *f, T2 *fb,
+                      mfem::AssemblyLevel assembly_level, int skip_zeros,
+                      bool no_assembly = false)
+{
+  auto a = std::make_unique<mfem::SymmetricBilinearForm>(&fespace);
+  if (f && !f->empty())
+  {
+    a->AddDomainIntegrator(new mfem::MixedGradGradIntegrator(*f));
+  }
+  if (fb && !fb->empty())
+  {
+    a->AddBoundaryIntegrator(new mfem::MixedGradGradIntegrator(*fb));
+  }
+  if (!no_assembly)
+  {
+    a->SetAssemblyLevel(assembly_level);
+    a->Assemble(skip_zeros);
+    a->Finalize(skip_zeros);
+  }
+  return a;
+}
+
+}  // namespace
+
+std::unique_ptr<Operator>
+SpaceOperator::GetStiffnessMatrix(Operator::DiagonalPolicy diag_policy)
```

Comment:
I agree that at this point `Get*Operator` would be a more appropriate name than `Get*Matrix`. But this would introduce a fairly large amount of change for the rename so I prefer to hold off on this for now. Users can think of the returned object as a matrix so the naming isn't totally wrong.